text/0000-interior.md

```diff
@@ -0,0 +1,348 @@
+[proposed linear-types facility][lineartypes] to allow wrapping a
+linear-type in an affine type (which can be useful to make linear
+types more ergonomic in some circumstances). In the linear types
+proposal, implementing `Drop` against a compound type containing a
+linear field will allow the compound container type to be treated as
+affine. In that case, the linear field must be moved out during the
+drop-hook to satisfy linearity constraints, but the current drop-hook
+API does not allow partial moves from the dropped value. The
+drop-behavior defined in this RFC *would* be forward-compatible with
+the proposed linear-types mechanism.
+
+[lineartypes]: https://github.com/rust-lang/rfcs/issues/814
+
+This RFC proposes that the drop-hook take a form of `self` by value,
+instead of by reference. I say "a form of", because we want the
+drop-hook to allow moves out of the `self` argument, but partial moves
+are disallowed for types implementing `Drop`. Also, if the argument
+were the direct `self` value, then it would usually imply that new
+drop-glue would be inserted at the end of the drop-hook, resulting in
+an infinite recursive loop at the end of the drop-hook. What this
+approach needs is a different view of the `self` argument to `drop`
+which does *not* implement `Drop`, so that partial moves would be
+allowed, and so that the drop-hook would not re-invoke itself
+recursively by default.
+
+Other solutions to this issue in current-Rust require circumlocutions
+for a `Drop`-implementing compound type to control whether or how its
+embedded fields get dropped. These approaches will be discussed in the
+`Alternatives` section, below.
+
+# Detailed design
+
+We define an `Interior<T>` type and associated `"interior"` language
+item:
+
+```rust
+#[lang="interior"]
+struct Interior<T>(T);
+```
+
+An instance of `Interior<T>` is defined to be structurally identical
+to `T`, but is an independent type. Since it is an independent type,
+it can have a unique set of `impl`s defined against it: the `impl`s
+defined against `T` are not inherited by `Interior<T>`. In particular,
+this means that `Interior<T>` does not implement `Drop`, which means
+it is (by default) valid to partially or fully deconstruct an
+`Interior<T>` instance (even where it would not be valid to
+deconstruct an instance of `T`).
+
+We also define a new `DropValue` trait to replace the legacy `Drop`
+trait, and which takes an instance of `Interior<Self>` by value:
+
+```rust
+#[lang="dropvalue"]
+trait DropValue {
+    fn drop_value(Interior<Self>);
+}
+```
+
+Define the following bare functions to convert between `T` and
+`Interior<T>`:
+
+```rust
+fn into_interior<T>(t: T) -> Interior<T>;
+fn from_interior<T>(t: Interior<T>) -> T;
+fn as_interior<T>(t: &T) -> &Interior<T>;
+fn as_interior_mut<T>(t: &mut T) -> &mut Interior<T>;
+fn of_interior<T>(t: &Interior<T>) -> &T;
+fn of_interior_mut<T>(t: &mut Interior<T>) -> &mut T;
```

Comment:
Consider an example like this:

```rust
struct MaybeCrashOnDrop { c: bool }
impl Drop for MaybeCrashOnDrop {
    fn drop(&mut self) {
        if self.c {
            unsafe { *(1 as *mut u8) = 0 }
        }
    }
}
pub struct InteriorUnsafe { m: MaybeCrashOnDrop }
impl InteriorUnsafe {
    pub fn new() -> InteriorUnsafe {
        InteriorUnsafe { m: MaybeCrashOnDrop{ c: true } }
    }
}
impl Drop for InteriorUnsafe {
    fn drop(&mut self) {
        self.m.c = false;
    }
}
```
Currently, safe code can't break safety rules using an InteriorUnsafe, but it could if `into_interior` were considered safe, I think?